plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/performance/MapOrSetKeyInspection.java

```diff
@@ -0,0 +1,191 @@
+// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package com.siyeh.ig.performance;
+
+import com.intellij.psi.*;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.siyeh.ig.BaseInspection;
+import com.siyeh.ig.BaseInspectionVisitor;
+import org.jetbrains.annotations.NonNls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class MapOrSetKeyInspection extends BaseInspection {
+
+  protected abstract boolean shouldTriggerOnKeyType(PsiType argumentType);
+
+  @Override
+  public final BaseInspectionVisitor buildVisitor() {
+    return new ContainsEntryOfInterestVisitor();
+  }
+
+  private static ClassType getClassType(@Nullable PsiClass aClass) {
+    return isMapOrSet(aClass, new HashSet<>());
+  }
+
+  private static ClassType isMapOrSet(
+    @Nullable PsiClass aClass, Set<? super PsiClass> visitedClasses) {
+    if (aClass == null) {
+      return ClassType.OTHER;
+    }
+    if (!visitedClasses.add(aClass)) {
+      return ClassType.OTHER;
+    }
+    @NonNls final String className = aClass.getQualifiedName();
+    if (CommonClassNames.JAVA_UTIL_SET.equals(className)) {
+      return ClassType.SET;
+    }
+    if (CommonClassNames.JAVA_UTIL_MAP.equals(className)) {
+      return ClassType.MAP;
+    }
+    final PsiClass[] supers = aClass.getSupers();
```

Comment:
It doesn't cover 2 levels of inheritance, like `MyMap extends AbstractMap`. I see that it was like that but still